### PR TITLE
filters optimization in `ranges`

### PR DESCRIPTION
**Pull request by gouhongshen**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #https://github.com/matrixorigin/matrixone/issues/14517

## What this PR does / why we need it:
1. pre-compile to reduce compile latency.
2. support `in`, `prefix in`, `between`, `prefix_between`, `isnotnull`, `isnull`, `>=`, `=`, `<=`, `or` and `and`.
3. simplified codes.
4. quick break iteration and binary search to find blk.

### CODE REVIEW

**Review by LeftHandCold — APPROVED** [review]

**Review by ouyuanning — APPROVED** [review]

File: pkg/vm/engine/disttae/txn_table.go [hunk 1]
```diff
@@ -663,80 +657,31 @@ func (tbl *txnTable) rangesOnePart(
...
-		ForeachBlkInObjStatsList(true, nil, func(blk objectio.BlockInfo, _ objectio.BlockObject) bool {
-			if tbl.db.txn.hasUncommittedDeletesOnBlock(&blk.BlockID) {
-				dirtyBlks[blk.BlockID] = struct{}{}
-			}
-			return true
-		}, uncommittedObjects...)
+	// for dynamic parameter, substitute param ref and const fold cast expression here to improve performance
+	newExprs, err := plan2.ConstandFoldList(exprs, tbl.proc.Load(), true)
+	if err == nil {
+		exprs = newExprs
```
Comment (ouyuanning) [inline pkg/vm/engine/disttae/txn_table.go:682]: 好像这里还是不能替换掉。你可以验证一下。

File: pkg/vm/engine/disttae/txn_table.go [hunk 2]
```diff
@@ -567,21 +567,21 @@ func (tbl *txnTable) Ranges(ctx context.Context, exprs []*plan.Expr) (ranges eng
...
+	//newExprs := make([]*plan.Expr, len(exprs))
 	for i := range exprs {
-		newExprs[i] = plan2.DeepCopyExpr(exprs[i])
+		//newExprs[i] = plan2.DeepCopyExpr(exprs[i])
 		// newExprs[i] = plan2.SubstitueParam(newExprs[i], tbl.proc)
-		foldedExpr, _ := plan2.ConstantFold(batch.EmptyForConstFoldBatch, newExprs[i], tbl.proc.Load(), true)
+		foldedExpr, _ := plan2.ConstantFold(batch.EmptyForConstFoldBatch, exprs[i], tbl.proc.Load(), true)
 		if foldedExpr != nil {
-			newExprs[i] = foldedExpr
+			exprs[i] = foldedExpr
```
[the remark this answers is not in the document]
Comment (aunjgr) [reply pkg/vm/engine/disttae/txn_table.go]: 确实需要确认一下，prepare语句不能修改plan里的表达式

[the remark this answers is not in the document]
Comment (aunjgr) [reply pkg/vm/engine/disttae/txn_table.go]: 我试了一下，查询类语句prepare不会改plan内容

**Review by aunjgr — APPROVED** [review]

File: pkg/vm/engine/disttae/txn_table.go (hunk 1 above)
Comment (badboynt1) [reply pkg/vm/engine/disttae/txn_table.go:682]: ConstandFoldList这里做过copy的，应该没有问题

File: pkg/vm/engine/disttae/txn_table.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (badboynt1) [reply pkg/vm/engine/disttae/txn_table.go]: 这是哪里 我怎么没找到

**Review by badboynt1 — APPROVED** [review]

**Pull request merged**